libs/discord/discord.cpp

```diff
@@ -0,0 +1,212 @@
+static int old_team = -1;
+static int old_match_state = -1;
+static int old_round = -1;
+
+
+void DiscordFrame()
+{
+	ZoneScoped;
+
+	if (app.activities)
+	{
+		if( (cls.state == CA_CONNECTED || cls.state == CA_ACTIVE) && client_gs.module )
+		{
+
+			if (old_state != cls.state ||
+				old_team != cg.predictedPlayerState.real_team ||
+				old_match_state != client_gs.gameState.match_state ||
+				old_round != client_gs.gameState.round_num)
+			{
+				old_state = cls.state;
+				old_team = cg.predictedPlayerState.real_team;
+				old_match_state = client_gs.gameState.match_state;
+				old_round = client_gs.gameState.round_num;
+
+				DiscordActivity activity = {};
+				//activity.type = DiscordActivityType_Playing;
+
+				if (cg.predictedPlayerState.real_team == TEAM_SPECTATOR) {
+					Q_strncpyz(activity.details, "SPECTATING");
+				} else if (client_gs.gameState.match_state <= MatchState_Warmup) {
+					Q_strncpyz(activity.details, "WARMUP");
+				} else if (client_gs.gameState.match_state <= MatchState_Playing) {
+					ggformat(activity.details, "ROUND {}", client_gs.gameState.round_num);
+				}
+
+				bool is_bomb = GS_TeamBasedGametype( &client_gs );
+				if (is_bomb) {
+					auto& alpha = client_gs.gameState.teams[ TEAM_ALPHA ];
+					auto& beta = client_gs.gameState.teams[ TEAM_BETA ];
+					ggformat(activity.state, "bomb: {}-{}", alpha.score, beta.score );
+				} else {
+					Q_strncpyz(activity.state, "gladiator" );
+				}
+
+				if (client_gs.gameState.match_state_start_time) {
+					time_t now = time( NULL );
+					activity.timestamps.start = now - client_gs.gameState.match_state_start_time / 1000;
+				}
+
+				if (cl.map && !Q_strnicmp(cl.map->name, "maps/", 5))
+				{
+					char tmpmap[128];
+					ggformat(tmpmap, cl.map->name + 5);
+					COM_StripExtension(tmpmap);
+					ggformat(activity.assets.large_image, "map-{}", tmpmap);
+					ggformat(activity.assets.large_text, "Playing on map {}", tmpmap);
+				}
+				auto gt = is_bomb ? "bomb" : "gladiator";
+				ggformat(activity.assets.small_image, "gt-{}", is_bomb ? "bomb" : "gladiator");
+				ggformat(activity.assets.small_text, "Gametype: {}", gt);
+				activity.instance = 1;
+
+				Com_GGPrint("Discord update: {}, {}, {}, {}", activity.details, activity.state, activity.assets.large_text, activity.assets.small_text);
+				app.activities->update_activity(app.activities, &activity, &app, UpdateActivityCallback);
+			}
+		}
+		else if (old_state != CA_DISCONNECTED) {
+			old_state = CA_DISCONNECTED;
+			old_team = -1;
+			old_match_state = -1;
+			old_round = -1;
+
+			DiscordActivity activity = {};
+			Q_strncpyz(activity.details, "MENU");
+			Q_strncpyz(activity.assets.large_image, "mainmenu");
+
+			Com_GGPrint("Discord update: {}", activity.details);
+			app.activities->update_activity(app.activities, &activity, &app, UpdateActivityCallback);
+		}
+	}
+
+
+	if (app.core)
+	{
+		EDiscordResult result = app.core->run_callbacks(app.core);
```

Comment:
This needs to run regularly, but I guess this could run less often,
is there something that runs on a slower tick than every frame?

Reply:
There's some ad-hoc bits that do like `static int last_second;` to run once per second, don't have anything proper